### DISCUSSION TITLE
카프카(kafka) / msk 지표는 무엇이 있을까?

### DISCUSSION THREAD

**Discussion by kyungjunleeme** [kyungjunleeme/kafka_study#42] [Q&A]
https://docs.aws.amazon.com/ko_kr/msk/latest/developerguide/metrics-details.html

**Comment by hellices** [comment]
cloudwatch는 비용이 생각보다 많아서 규모가 커질수록 부담이 되더라구요.
저희는 grafana & prometheus를 자체 구성해서 kafka metric도 별도로 받아서 운영하고 있습니다.
strimzi에서 제공하는 모니터링 지표가 생각보다 구체적이어서 많은 도움을 받고 있습니다.
https://strimzi.io/docs/operators/latest/overview#metrics-overview_str

**Comment by kyungjunleeme** [comment]
cf) 

안녕하세요! 질문하신 “카프카 크래프트 모드”에서 모니터링할 만한 항목에 대해 말씀드리겠습니다. “카프카 크래프트 모드”라는 표현이 Apache Kafka의 KRaft (Kafka Raft Consensus) 모드를 의미하는 것으로 보이는데, 이는 Kafka에서 ZooKeeper 의존성을 제거하고 자체 Raft 기반 합의 메커니즘을 사용하는 새로운 아키텍처를 말합니다. KRaft 모드에서 모니터링해야 할 주요 항목은 기존 ZooKeeper 기반 모니터링과 일부 유사하지만, 합의 메커니즘과 관련된 특수한 메트릭도 포함됩니다. 아래에 주요 모니터링 항목을 정리해 드리겠습니다.

1. 클러스터 상태 모니터링
	•	컨트롤러 상태 (Active Controller Count) KRaft 모드에서는 ZooKeeper 대신 컨트롤러 노드가 클러스터 메타데이터를 관리합니다. 항상 하나의 활성 컨트롤러만 존재해야 하며, kafka.controller:type=KafkaController,name=ActiveControllerCount 메트릭을 통해 확인할 수 있습니다. 값이 1이 아닌 경우 문제가 있을 수 있습니다.
	•	컨트롤러 가용성 컨트롤러 노드의 업타임과 장애 발생 여부를 모니터링하세요. 컨트롤러가 자주 바뀌거나 다운되면 클러스터 안정성에 영향을 줄 수 있습니다.
	•	파티션 리더 분포 리더 파티션이 특정 브로커에 치우치지 않도록 분포를 체크하세요. UnderReplicatedPartitions가 증가하면 복제 문제가 있을 수 있습니다.

2. Raft 합의 메커니즘 관련 메트릭
	•	Raft 리더 선출 시간 (Leader Election Latency) KRaft는 Raft 프로토콜을 사용하므로 리더 선출이 원활히 이루어지는지 확인해야 합니다. 선출 시간이 길어지면 클러스터 응답성이 떨어질 수 있습니다.
	•	로그 복제 지연 (Replication Latency) Raft 로그 복제가 느려지면 데이터 일관성에 문제가 생길 수 있습니다. kafka.raft:replication-latency 같은 메트릭을 모니터링하세요.
	•	커밋 레이턴시 (Commit Latency) Raft에서 로그가 커밋되는 속도를 체크하여 합의 과정의 효율성을 평가할 수 있습니다.

3. 브로커 성능 모니터링
	•	네트워크 처리량 (Bytes In/Out) kafka.network:type=BrokerTopicMetrics,name=BytesInPerSec 및 BytesOutPerSec를 통해 브로커의 입출력 부하를 확인하세요.
	•	요청 처리 시간 (Request Latency) kafka.network:type=RequestMetrics,name=TotalTimeMs로 각 요청 타입(Produce, Fetch 등)의 지연 시간을 모니터링하세요.
	•	CPU 및 메모리 사용량 KRaft 모드에서 컨트롤러와 브로커가 동일한 프로세스에서 실행될 수 있으므로 리소스 사용량이 증가할 가능성이 있습니다.

4. 토픽 및 파티션 상태
	•	오프라인 파티션 수 (Offline Partitions) kafka.controller:type=ControllerStats,name=OfflinePartitionsCount가 0이 아니면 파티션이 비정상 상태일 수 있습니다.
	•	언더 리플리케이티드 파티션 (UnderReplicatedPartitions) kafka.server:type=ReplicaManager,name=UnderReplicatedPartitions를 통해 복제가 제대로 안 되는 파티션을 감지하세요.
	•	메시지 처리율 토픽별 Produce/Consume 속도를 모니터링하여 애플리케이션 성능을 평가합니다.

5. 프로듀서와 컨슈머 모니터링
	•	컨슈머 랙 (Consumer Lag) kafka.consumer:type=ConsumerFetcherManager,name=MaxLag를 통해 컨슈머가 토픽의 최신 메시지를 얼마나 뒤처져 처리하는지 확인하세요.
	•	프로듀서 오류율 메시지 전송 실패율(kafka.producer:type=producer-metrics,name=request-error-rate)을 체크하여 프로듀서 문제를 조기에 감지하세요.

6. 디스크 및 로그 상태
	•	로그 세그먼트 크기 및 개수 kafka.log:type=Log,name=Size로 디스크 사용량을 모니터링하고, 로그 세그먼트가 과도하게 커지지 않도록 주의하세요.
	•	디스크 I/O KRaft 모드에서도 디스크 성능이 중요하므로 읽기/쓰기 속도와 대기 시간을 확인하세요.

모니터링 도구 추천
	•	Prometheus + Grafana: JMX Exporter를 통해 Kafka 메트릭을 수집하고, Grafana로 시각화.
	•	Kafka UI: 토픽, 브로커, 컨슈머 그룹 상태를 직관적으로 확인 가능.
	•	Burrow: 컨슈머 랙 모니터링에 특화된 도구로, KRaft 모드에서도 유용.

추가 팁
KRaft는 ZooKeeper를 제거하면서 설정이 간소화되었지만, 컨트롤러 노드의 역할이 커졌습니다. 따라서 컨트롤러의 안정성과 Raft 합의 과정에 특히 주의를 기울이는 것이 중요합니다. 모니터링 항목을 설정할 때는 운영 환경(예: 트래픽 규모, 브로커 수)에 맞게 우선순위를 정하세요.
궁금한 점이 더 있으면 언제든 물어보세요! 도움이 되기를 바랍니다.
